api/services/container/container.proto

```diff
@@ -0,0 +1,79 @@
+syntax = "proto3";
+
+package containerd.v1;
+
+import "gogoproto/gogo.proto";
+import "google/protobuf/empty.proto";
+import "google/protobuf/field_mask.proto";
+import "github.com/containerd/containerd/api/types/mount/mount.proto";
+import "github.com/containerd/containerd/api/types/descriptor/descriptor.proto";
+
+// Containers provides metadata storage for containers used in the execution
+// service.
+//
+// The objects here provide an state-independent view of containers for use in
+// management and resource pinning. From that perspective, contaienrs do not
+// have a "state" but rather this is the set of resources that will be 
+//
+// From the perspective of the execution service, these objects represent the
+// base parameters for creating a container process.
+//
+// In general, when looking to add fields for this type, first ask yourself
+// whether or not the function of the field has to do with runtime exeuction or
+// is invariant of the runtime state of the container. If it has to do with
+// runtime, or changes as the "container" is started and stops, it probably
+// doesn't belong on this object.
+service Containers {
+	Get(GetContainerRequest) returns (GetContainerResponse);
+	List(ListContainersRequest) returns (stream ListContainersResponse);
+	Create(CreateContainerRequest) returns (CreateContainerResponse);
+	Update(UpdateContainerRequest) returns (UpdateContainerResponse);
+	Delete(DeleteContainerRequest) returns (google.protobuf.Empty);
+}
+
+message Container {
+	// ID is the user-specified identifier.
+	//
+	// This field may not be updated.
+	string id = 1 [(gogoproto.customname) = "ID"];
+
+	// Labels provides an area to include arbitrary data on containers.
+	//
+	// Note that to add a new value to this field, read the existing set and
+	// include the entire result in the update call.
+	map<string, string> labels  = 2;
+
+	// Runtime specifies which runtime to use for executing this container.
+	string runtime = 3;
+
+	// Spec, stored in the content store, to be used when creating the
+	// container. This is usually system specific.
+	//
+	// Objects specified in this field will not be garbage collected.
+	types.Descriptor spec = 4;
+
+	// Snapshots provides a list of snapshots to use for the this container.
+	//
+	// Order matters. The mounts corresponding to these snapshots should be
+	// provided as part of the execution service task creation in the same
+	// order in this field.
+	//
+	// Snapshots referenced in this field will not be garbage collected.
+	repeated string snapshots = 5;
+}
+
+// UpdateContainerRequest updates the metadata on one or more container.
+// 
+// The operation should follow semantics described in
+// https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/field-mask,
+// unless otherwise qualified.
+message UpdateContainerRequest {
```

Comment:
What is the expected use case for `update` if the metadata stored is expected to be `state-independent`?